### PR TITLE
Example of use differential evolution function

### PR DESCRIPTION
**Pull request by JudytaBakowska**

### CODE REVIEW

File: calculate_error.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import numpy as np
```
Comment (tunczyk101) [inline calculate_error.py:1]: Dodałabym import:
`from sklearn.metrics import mean_absolute_error, mean_squared_error`

i potem już tego używała w głównej funkcji.
I jeżeli tak to usunęłabym poniższe implementacje.

File: calculate_error.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+def mean_absolute_error(y_true, y_pred):
```
Comment (tunczyk101) [inline calculate_error.py]: Jeżeli wyższy komentarz to: Można usunąć

File: calculate_error.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Calculates the mean absolute error (MAE) between the actual y_true values and the predicted y_pred values.
+    :param y_true: real values
+    :param y_pred: predicted values
+    :return: mean absolute error
+    """
+    mae = np.mean(np.abs(y_true - y_pred))
+    return mae
+
+def mean_squared_error(y_true, y_pred):
```
Comment (tunczyk101) [inline calculate_error.py]: Jeżeli wyższy komentarz to: Można usunąć

File: calculate_error.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    values of y_true and the predicted values of y_pred.
+    :param y_true: real values
+    :param y_pred: predicted values
+    
+    :return: mse - mean square error
+    """
+    mse = np.mean((y_true - y_pred) ** 2)
+    return mse
+
+def root_mean_squared_error(y_true, y_pred):
```
Comment (tunczyk101) [inline calculate_error.py]: Jeżeli wyższy komentarz to: Można usunąć

File: calculate_error.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    :param *args: predicted values
+    :param metric: name of the metric ('mae', 'rmse' or 'mse')
+    :return: error -- the value of the error metric
+    """
+    true_val = np.array([1,2,3,4,5,6])
+    
+    if metric == 'mae':
+        error = mean_absolute_error(true_val,args)
+    elif metric == 'rmse':
+        error =  root_mean_squared_error(true_val,args)
```
Comment (tunczyk101) [inline calculate_error.py]: Jeżeli wyższy komentarz to: tylko w tym `elif` zmiana: mean_squared_error(true_val, args, squared=False)

File: calculate_error.py (hunk 1 above)
Comment (tunczyk101) [inline calculate_error.py:1]: Komentarz do całości pliku -> formatowanie
Brak spacji po przecinkach, ale odstępy między funkcjami przepisowo 👍 

Jak nie chce ci się samej (zupełnie zrozumiałe ja np. nie formatuję sama plików) to:
w terminalu `pip install black` + dla notebooków `pip install jupyter-black jupyter`
I teraz przed każdym Commitem:
`black {path_to_file_to_format}`

File: calculate_error.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Calculates the selected error metric from the actual true_values 
+    and predicted args values.
+    :param *args: predicted values
+    :param metric: name of the metric ('mae', 'rmse' or 'mse')
+    :return: error -- the value of the error metric
+    """
+    true_val = np.array([1,2,3,4,5,6])
+    
+    if metric == 'mae':
```
Comment (tunczyk101) [inline calculate_error.py]: Nadal zachęcam do rozważenia `match` + `case` oraz dyskusji w kometarzu jeżeli się nie zgadzasz.

File: calculate_error.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    rmse = np.sqrt(mean_squared_error(y_true,y_pred))
+    return rmse
+
+
+def calculate_error(args, metric='mae'):
+    """
+    Calculates the selected error metric from the actual true_values 
+    and predicted args values.
+    :param *args: predicted values
```
Comment (tunczyk101) [inline calculate_error.py]: :param *args: -> :param args:

File: calculate_error.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    mae = np.mean(np.abs(y_true - y_pred))
+    return mae
+
+def mean_squared_error(y_true, y_pred):
+    """
+    Calculates the mean square error (MSE) between the actual
+    values of y_true and the predicted values of y_pred.
+    :param y_true: real values
+    :param y_pred: predicted values
+
```
Comment (tunczyk101) [inline calculate_error.py]: Do usunięcia.

File: example_of_use_differential_evolution.ipynb [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    "##### example parameters"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "bounds = [[-4,4]] #example bounds\n",
```
Comment (tunczyk101) [inline example_of_use_differential_evolution.ipynb]: Szukamy tutaj ciągu o długości `len(true_val)`, więc `bounds` też musi być takiej długości. To działa rzecz jasna, ale dlatego, że `numpy` jest mądry i uznaje, że odejmujemy stałą od wszystkich wartości w tablicy. // Dla sklearn to by nie działało, ale znalazłabyś "błąd"(?) 

Jeżeli to działanie było zamierzone i celem było znalezienie liczby, która ma jak najmniejszą średnią odległość od danych punktów `true_val` to oczywiście zwracam honor. 🙏

File: example_of_use_differential_evolution.ipynb [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.12.0"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
Comment (tunczyk101) [inline example_of_use_differential_evolution.ipynb:161]: Spróbuj jeszcze przekazać w wywołaniu `differential_evolution` `metric="mse"`
`differential_evolution(calculate_error, bounds, mutation = mutation,recombination = recombination, metric="mse")`

Nie wiem, czy zadziała, ale może...

**Review by tunczyk101 — CHANGES_REQUESTED** [review]
Jakieś głupoty i głównie sugestie/niejasności odnośnie zamysłu co do danej implementacji.

**Comment by tunczyk101** [comment]
Komentarz co do rmse: Aktualizacja sklearn, brak parametru squared dla `mean_squared_error`, więc zamiast tego trzeba dodać w imporcie `root_mean_squared_error` i potem użyć

File: example_of_use_differential_evolution.ipynb (hunk 9 above)
Comment (JudytaBakowska) [reply example_of_use_differential_evolution.ipynb]: racja, poprawiłam bounds

**Comment by JudytaBakowska** [comment]
`differential_evolution(calculate_error, bounds, mutation = mutation,recombination = recombination, metric="mse")` - własnie w ten sposób nie działa, nie da się w ten sposób wrzucić metric do funkcji

Ogólnie teraz niby działa, bo w `best_predicted_values` pokazują się te przewidziane wartości
Ale niestety dalej mi się coś nie zgadza, bo właśnie nie da się w wywołaniu funkcji `differential_evolution` zmienić metryki z tej domyślej, a dodatkowo jak zmieniam `mutation`, `recombination` i `bounds`, to za każdym razem wychodzi ten sam wynik. Nie wiem czy czegoś mi brakuje w tym wywoływaniu, czy co jest nie tak, ale nie potrafię znaleźć rozwiązania tego problemu

File: calculate_error.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    :param metric: name of the metric ('mae', 'rmse' or 'mse')
+    :return: error -- the value of the error metric
+    """
+    true_val = np.array([1, 2, 3, 4, 5, 6])
+
+    match metric:
+        case "mae":
+            error = mean_absolute_error(true_val, args)
+        case "rmse":
+            error = root_mean_squared_error(true_val, args)
```
Comment (tunczyk101) [inline calculate_error.py:23]: ❤️

**Review by tunczyk101 — APPROVED** [review]
